pkg/webhook/cvc.go

```diff
@@ -0,0 +1,204 @@
+func isScalingInProgress(cvcObj *apis.CStorVolumeClaim) bool {
+	return len(cvcObj.Spec.Policy.ReplicaPoolInfo) != len(cvcObj.Status.PoolInfo)
+}
+
+// validateReplicaCount returns error if user modified the replica count after
+// provisioning the volume else return nil
+func validateReplicaCount(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	if cvcOldObj.Spec.ReplicaCount != cvcNewObj.Spec.ReplicaCount {
+		return errors.Errorf(
+			"cvc %s replicaCount got modified from %d to %d",
+			cvcOldObj.Name,
+			cvcOldObj.Spec.ReplicaCount,
+			cvcNewObj.Spec.ReplicaCount,
+		)
+	}
+	return nil
+}
+
+// validatePoolListChanges returns error if user modified existing pool names with new
+// pool name(s) or if user performed more than one replica scale down at a time
+func validatePoolListChanges(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	oldDesiredPoolNames := cvc.GetDesiredReplicaPoolNames(cvcOldObj)
+	newDesiredPoolNames := cvc.GetDesiredReplicaPoolNames(cvcNewObj)
+	modifiedPoolNames := util.ListDiff(oldDesiredPoolNames, newDesiredPoolNames)
+	if len(newDesiredPoolNames) >= len(oldDesiredPoolNames) {
+		// If no.of pools on new spec >= no.of pools in old spec(scaleup as well
+		// as migration then there all the pools in old spec must present in new
+		// spec)
+		if len(modifiedPoolNames) > 0 {
+			return errors.Errorf(
+				"volume replica migration directly by modifying pool names %v is not yet supported",
+				modifiedPoolNames,
+			)
+		}
+	} else {
+		// If no.of pools in new spec < no.of pools in old spec(scale down
+		// volume replica case) then there should at most one change in
+		// oldSpec.PoolInfo - newSpec.PoolInfo
+		if len(modifiedPoolNames) > 1 {
+			return errors.Errorf(
+				"Can't perform more than one replica scale down requested scale down count %d",
+				len(modifiedPoolNames),
+			)
+		}
+	}
+	return nil
+}
+
+// validateReplicaScaling returns error if user updated pool list when scaling is
+// already in progress.
+// Note: User can perform scaleup of multiple replicas by adding multiple pool
+//       names at time but not by updating CVC pool names with multiple edits.
+func validateReplicaScaling(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	if isScalingInProgress(cvcOldObj) {
+		// if old and new CVC has same count of pools then return true else
+		// return false
+		if len(cvcOldObj.Spec.Policy.ReplicaPoolInfo) != len(cvcNewObj.Spec.Policy.ReplicaPoolInfo) {
+			return errors.Errorf("scaling of CVC %s is already in progress", cvcOldObj.Name)
+		}
+	}
+	return nil
+}
+
+// validateStatusPoolList return error if content of pools under status doesn't
+// match with desired pool names(spec.poolNames)
+// NOTE: This validation is only to reject invalid changes of pool information
+// on CVC status.
+func validateStatusPoolList(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	replicaPoolNames := []string{}
+	if len(cvcOldObj.Spec.Policy.ReplicaPoolInfo) == 0 &&
```

Comment:
can we add validation that status should be `Bound` to allow any scaling operation to avoid this complex logic?

Reply:
No this is particularly to verify the content of status with the spec (https://github.com/openebs/maya/pull/1613#discussion_r383396155 to incorporate this kind of cases). IMO we shouldn't validate the status(because the status will be updated only by the controllers).